src/pybind/mgr/mds_autoscaler/module.py

```diff
@@ -0,0 +1,98 @@
+"""
+Automatically scale MDSs based on status of the file-system using the FSMap
+"""
+
+import errno
+import json
+import time
+import mgr_util
+import threading
+import uuid
+import logging
+from six import itervalues, iteritems
+from collections import defaultdict
+from prettytable import PrettyTable, PLAIN_COLUMNS
+from mgr_module import MgrModule
+
+import orchestrator
+
+log = logging.getLogger(__name__)
+
+"""
+Some terminology is made up for the purposes of this module:
+
+ - "grow/shrink" - increase/decrease the MDSs in the system
+"""
+
+class MDSAutoscaler(MgrModule):
+    """
+    MDS autoscaler.
+    """
+    NATIVE_OPTIONS = [
+        'max_mds'
+    ]
+
+    MODULE_OPTIONS = [
+        {
+            'name': 'sleep_interval',
+            'default': str(60),
+        },
+    ]
+
+    def __init__(self, *args, **kwargs):
+        super(MDSAutoscaler, self).__init__(*args, **kwargs)
+        self.fs_map = None
+        self.filesystems = None
+        self.standbys = None
+
+    def create_mds(self, fs_name):
+        spec = orchestrator.StatelessServiceSpec(fs_name)
+        try:
+            completion = self.add_mds(spec)
+            self._orchestrator_wait([completion])
+            orchestrator.raise_if_exception(completion)
+        except Exception as e:
+            # Don't let detailed orchestrator exceptions (python backtraces)
+            # bubble out to the user
+            self.log.exception(" Failed to create MDS daemons")
+            return -errno.EINVAL, "", str(e)
+        return 0, "", ""
+
+    def get_total_standby_count_required(self):
+        total = 0;
+        for fs in self.filesystems:
+            total += fs['mdsmap']['standby_count_wanted']
+        return total
+
+    def get_current_standby_count(self):
+        return len(self.standbys)
+
+    def verify_and_launch_mds(self):
+        standbys_required = self.get_total_standby_count_required()
+        standbys_current = self.get_current_standby_count()
+
+        self.log.info("standbys_required:{0}, standbys_current:{1}".format(standbys_required, standbys_current))
+        if standbys_required < standbys_current:
+            # delete some standbys
+            pass
+        if standbys_current < standbys_required:
+            current = standbys_current
+            while current != standbys_required:
+                self.log.info("launching standby mds ...")
+                self.create_mds(fs['default'])
+                current += 1
+                time.sleep(1)
+
+    def is_insufficient_standby(self, message):
+        return (message == 'Health check failed: insufficient standby MDS daemons available (MDS_INSUFFICIENT_STANDBY)')
```

Comment:
is such a comparison the only way to figure out if there are insufficient standbys? wouldn't `get_current_standby_count() < get_total_standby_count_required()` be a good enough?

Reply:
@vshankar you're right